tidy3d/components/medium.py

```diff
@@ -615,6 +615,56 @@ def _interp(
         interp_dataset = interp_dataset.where(interp_dataset <= max_val, max_val)
         return interp_dataset
 
+    def grids(self, bounds: Bound) -> Dict[str, Grid]:
+        """Make a :class:`.Grid` corresponding to the data in each ``eps_ii`` component.
+        The min and max coordinates along each dimension are bounded by ``bounds``."""
+
+        rmin, rmax = bounds
+        pt_mins = dict(zip("xyz", rmin))
+        pt_maxs = dict(zip("xyz", rmax))
+
+        def make_grid(scalar_field: ScalarFieldDataArray) -> Grid:
+            """Make a grid for a single dataset."""
+
+            def make_bound_coords(coords: np.ndarray, pt_min: float, pt_max: float) -> List[float]:
+                """Convert user supplied coords into boundary coords to use in :class:`.Grid`."""
+
+                # get coordinates of the bondaries halfway between user-supplied data
+                coord_bounds = (coords[1:] + coords[:1]) / 2.0
+
+                # filter out boundaries that lie outside geometry bounds
+                coord_bounds[coord_bounds <= pt_min] = pt_min
+                coord_bounds[coord_bounds >= pt_max] = pt_max
```

Comment:
Note that this doesn't technically filter out the coordinates; it truncates all coordinates that are out of bounds to the geometry bounds. So coord_bounds does not change size, but if you use this e.g. in volume integration, all of the elements that were snapped to the geometry boundary will have zero volume element.

Reply:
Yes, this is exactly what it's intended to do. The reason is that the vjp must return a Datastructure that is identical in form to the original `self` but with different values. With `0` volume elements for coords out of bounds, we ensure that those data points are not counted in the integral, but still show up in the final returned data structure. I can adjust the docstring as it's a relic of the first iteration where I actually filtered them out before realizing my mistake.

Reply:
Oh iiiiii seeee